Feature/map (네이버 지도 api 통신) 기능 완성

EverJun2 opened the pull request:
### ✅  client / **NaverMapClient** 클래스 생성
전반적으로 지도 api 와 통신하는 역할을 제공합니다


### ✅  dto / response / **GeocodeResponseDTO , ReverseGeocodeResponseDTO** 생성

> **GeocodeResponseDTO : 주소 -> 좌표 변환 요청 시 반환되는 dto 입니다.**
> .getAddresses().get(0).getX() 이런식으로 사용하면 됩니다
<br>

> **ReverseGeocodeResponseDTO : 좌표 -> 지역명 변환 요청 시 반환되는 dto 입니다**
>  **.getRegion().getArea1().getName()** -> 서울특별시, 경기도 이렇게 시,도를 반환해줍니다.
> **.getRegion().getArea2().getName()** -> 송파구, 수원시 이렇게 시군구를 반환해줍니다.

-----------------------------------

### 그 외 사항들
✅  **관련 테스트 코드 작성**
✅   **application.yml** 에 naver.map. 부분에 발급된 api 키와 암호를 입력하여 사용하도록 구성
✅  **config / Appconfig.class** -> resttemplete 등록 위해서 **추가**

**develop 브랜치 밑에서 작업해서 아마 post 랑 충돌없을 것 같습니다. 
post 완성되면 post 쪽에서 호출해서 사용하면 될것 같습니다!**

EverJun2 committed:
feat : Naver Map API 활용을 위한 application.yml 세팅

EverJun2 committed:
feat : Naver Map API 활용을 위한 DTO 기본 구조 추가 -> dto/request&response/map 내부

EverJun2 committed:
feat : Naver Map API 활용을 위한 MapController, MapService 인터페이스, MapServiceImpl 추가

EverJun2 committed:
refactor : 굳이 필요하지 않아보이는 DTO 삭제

EverJun2 committed:
feat : MapService 인터페이스 작성

EverJun2 committed:
refactor : dto 단순화 및 그에 따른 수정

EverJun2 committed:
feat : MapController 작성

EverJun2 committed:
feat : AppConfig 작성 -> RestTemplate 빈 등록

EverJun2 committed:
refactor : application.yml 에서 naver map url 부분 삭제

EverJun2 committed:
feat : NaverMapClient 클래스 기본 구조 작성 (지도 API 와 직접적 통신)

EverJun2 committed:
feat : 주소 검색 클래스 작성 (NaverMapClient.class)

EverJun2 committed:
refactor : Map 관련 controller, service 삭제

EverJun2 committed:
feat : 네이버API 호출 시 response 용 DTO 작성 (GeocodeResponse)

EverJun2 committed:
feat : 네이버 map api 호출 테스트 코드 작성

EverJun2 committed:
feat : 네이버지도 api 를 활용한 주소 -> 좌표 변경 기능 완성 (callGeocodingAPI () )

EverJun2 committed:
feat : ReverseGeocodeResponseDTO 추가 / 좌표 -> 지역명 변환 기능 추가

EverJun2 committed:
refactor : GeocodeResponse.java -> GeocodeResponseDTO.java 이름 변경

EverJun2 committed:
feat : 좌표 -> 주소 변환 기능 추가

EverJun2 committed:
feat : 좌표 -> 주소 변환 기능 테스트 코드 작성

EverJun2 committed:
refactor : GeocodeDTO.java 삭제

EverJun2 committed:
comment : GeocodeResponseDTO javadoc 작성

EverJun2 committed:
comment : ReverseGeocodeResponseDTO javadoc 작성

EverJun2 committed:
comment : NaverMapClient javadoc 작성

whrjsgml0000 commented on src/main/java/club/inq/team1/client/NaverMapClient.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Value("${naver.map.client-secret}")
+    private String client_secret;
+
```
clientSecret 이런 식으로 변수명 변경하는게 좋을 것 같아요

whrjsgml0000 replied:
```java
@Value
```
어노테이션을 이용해서 유지보수에 용이할 것 같아요!

whrjsgml0000 commented on src/main/java/club/inq/team1/client/NaverMapClient.java:
```diff
@@ -0,0 +1,99 @@
+            GeocodeResponseDTO.class
+        );
+
+        return response.getBody();
```
response에서 응답 코드를 확인 후 변환이 똑바로 일어나지 않았을 경우 예외처리하는 로직이 있으면 좋을 것 같아요

whrjsgml0000 commented on src/main/java/club/inq/team1/client/NaverMapClient.java:
```diff
@@ -0,0 +1,99 @@
+     * @return GeocodeResponseDTO
+     */
+    public GeocodeResponseDTO callGeocodingAPI(String address) {
+        String url = "https://naveropenapi.apigw.ntruss.com/map-geocode/v2/geocode";
```
이 부분은 해당 클래스 상단에 
```java
private final String GEOCODE = "..."; 
```
이런 식으로 작성하면 좋을 것 같아요!

whrjsgml0000 commented on src/test/java/club/inq/team1/client/NaverMapClientTest.java:
```diff
@@ -0,0 +1,64 @@
+    GeocodeResponseDTO geocodeResponseDTO = new GeocodeResponseDTO();
+
+    @Autowired
+    ReverseGeocodeResponseDTO reverseGeocodeResponseDTO = new ReverseGeocodeResponseDTO();
```
DTO에 Autowired는 없어도 될 것 같아요!
또 테스트 코드 상단에 따로 new를 이용해서 만들지 않아도 될 것 같아요

whrjsgml0000 commented on src/main/java/club/inq/team1/dto/response/ReverseGeocodeResponseDTO.java:
```diff
@@ -0,0 +1,66 @@
+ */
+
+@Getter
+@Component
```
```java
@Component
``` 
어노테이션은 없어도 될 것 같아요

whrjsgml0000 reviewed (changes requested):
전체적으로 코드를 깔끔하게 작성해주셨네요!

NaverMapClient 클래스에서 실제로 담당하게 될 역할은 네이버 지도 api를 이용해서 주소의 좌표 추출과 좌표의 주소 추출이라는 점에서 NaverMapService 등으로 이름을 바꾸는 것도 좋을 것 같아요!

API 찾아보고 코드 작성하느라 고생 많으셨어요!

EverJun2 committed:
refactor : NaverMapClient -> NaverMapService 이름 변경

EverJun2 committed:
refactor : client_id, client_secret -> clientId, clientSecret 으로 수정

EverJun2 committed:
refactor : url을 전역 변수로 수정

EverJun2 committed:
refactor : test 코드 수정

EverJun2 committed:
refactor : ReverseGeocodeResponseDTO @component 애노테이션 삭제

EverJun2 committed:
refactor : NaverMapService 에서 response 가 200이 아닌 경우 null 을 반환하도록 수정

whrjsgml0000 commented on src/main/java/club/inq/team1/dto/response/ReverseGeocodeResponseDTO.java:
```diff
@@ -0,0 +1,65 @@
+        private Area area1;  // 예: 서울, 경기도 (도시)
+
+        @JsonProperty("area2")
+        private Area area2;  // 예: 송파구, 수원시 (시/구/군)
```
이 부분에 JsonProperty 어노테이션을 적용했으니까 데이터를 바인딩하는데 있어서 문제는 없을거에요.
필드 이름을 적절하게 바꾸는 것이 사용에 있어서 편리할 것 같아요

whrjsgml0000 commented on src/test/java/club/inq/team1/client/NaverMapClientTest.java:
```diff
@@ -0,0 +1,61 @@
+class NaverMapClientTest {
+
+    @Autowired
+    NaverMapService naverMapClient = new NaverMapService();
```
Autowired를 이용할 때는 new 를 통해 주입을 하지 않고 필드만 만들어 놓으면 될거에요.

whrjsgml0000 commented on src/main/java/club/inq/team1/client/NaverMapService.java:
```diff
@@ -0,0 +1,106 @@
+ * Naver Map API 와 직접적으로 통신하는 클래스입니다
+ */
+
+@Component
```
Component 대신 Service로 해도 될 것 같아요

whrjsgml0000 commented on src/main/java/club/inq/team1/client/NaverMapService.java:
```diff
@@ -0,0 +1,106 @@
+ */
+
+@Component
+public class NaverMapService {
```
현재 NaverMapService를 직접적으로 이용하고 있는데 MapService 인터페이스를 하나 만든 다음 상속받아서 사용하도록 만들면 좋을 것 같아요!

whrjsgml0000 reviewed (changes requested):
고생 많으셨어요! 제 생각에는 조금만 더 손보면 나중에 신경을 안써도 될 것 같아서 한번 더 디테일하게 코드 수정을 요청하게 됐네요... 이것만 수정하시고 난 다음 제가 게시글 관련 부분 작업이 끝나면 거기서 다시 역할분배를 하도록 할게요!

EverJun2 committed:
refactor : NaverMapService @component -> @service 로 수정

EverJun2 committed:
refactor : NaverMapServiceTest 코드 수정

EverJun2 committed:
refactor : ReverseGeocodeResponseDTO 필드 이름수정
1. area0 -> country (국가코드)
2. area1 -> sido (시도, 서울특별시, 경기도 등)
3. area2 -> sigugun (시구군, 수원시, 송파구, 칠곡군 등)

EverJun2 committed:
refactor : ReverseGeocodeResponseDTO 필드 이름수정에 따른 테스트코드 수정

EverJun2 committed:
refactor : ReverseGeocodeResponseDTO 필드 이름수정에 따른 테스트코드 수정

EverJun2 committed:
feat : MapService 인터페이스 작성

EverJun2 committed:
refactor : MapService 인터페이스 작성에 따른 수정 & 기존 header 생성 메소드를 한개로 통합

EverJun2 commented:
NaverMapService 파일을 service 디렉토리 안으로 이동할까요??

whrjsgml0000 commented:
> NaverMapService 파일을 service 디렉토리 안으로 이동할까요??

service 안에 impl 디렉토리로 옮기시면 될 것 같아요
명명도 NaverMapServiceImpl 이런 식으로 하시면 더 좋을 것 같고요
MapService는 service 디렉토리로 옮기시면 될 것 같아요

whrjsgml0000 reviewed (approved):
LGTM!

EverJun2 committed:
refactor : MapService 및 NaverMapServiceImpl 디렉토리 수정
